File: 0069_Sqrt(x)/Solution2.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int mySqrt(int x) {
        if (x == 0) return 0;
        long long len = x;
        while (len * len > x) {
            len = (len + x / len) / 2;
        }
        return len;
    }
};


void test(const string& test_name,
          int x,
          int expected)
{
    Solution s;
    int res = s.mySqrt(x);
    if (res == expected) {
        cout << test_name << " succeed" << endl;
    } else {
        cout << test_name << " fail" << endl;
    }
}

int main()
{
    int x1 = 4;
    int expected1 = 2;
    test("test1", x1, expected1);

    int x2 = 8;
    int expected2 = 2;
    test("test2", x2, expected2);

    int x3 = 2147395600;
    int expected3 = 46340;
    test("test3", x3, expected3);

    return 0;
}

File: 0069_Sqrt(x)/Solution2.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x <= 1:
            return x

        r = x
        while r * r > x:
            r = (r + x / r) // 2
        return int(r)


def test(test_name, x, expected):
    res = Solution().mySqrt(x)
    if isinstance(res, int) and res == expected:
        print(test_name + ' success.')
    else:
        print(test_name + ' failed.')


if __name__ == "__main__":
    x1, expected1 = 4, 2
    test("test1", x1, expected1)

    x2, expected2 = 8, 2
    test('test2', x2, expected2)

    x3, expected3 = 2, 1
    test('test3', x3, expected3)

    x4, expected4 = 4, 2
    test('test4', x4, expected4)

    x5, expected5 = 0, 0
    test('test5', x5, expected5)
